resources/assets/lib/store.ts

```diff
@@ -0,0 +1,167 @@
+ */
+
+import Shopify from 'shopify-buy';
+import { toShopifyVariantId } from 'shopify-gid';
+
+declare global {
+  interface Window {
+    Store: Store;
+  }
+}
+
+// process.env.$ has to be static as it is injected by webpack at compile time.
+const options = {
+  domain: process.env.SHOPIFY_DOMAIN,
+  storefrontAccessToken: process.env.SHOPIFY_STOREFRONT_TOKEN,
+};
+
+const client = Shopify.buildClient(options);
+
+interface LineItem {
+  quantity: number;
+  variantId: string;
+}
+
+export class Store {
+  private static instance: Store;
+
+  static init() {
+    if (this.instance == null) {
+      this.instance = new Store();
+    }
+
+    return this.instance;
+  }
+
+  private constructor() {
+    $(document).on('click', '.js-store-checkout', this.beginCheckout.bind(this));
+    $(document).on('click', '.js-store-resume-checkout', this.resumeCheckout.bind(this));
+
+    $(document).on('turbolinks:load', () => {
+      $('.js-store-checkout').prop('disabled', false);
+    });
+
+    $('.js-store-checkout').prop('disabled', false);
+  }
+
+  async beginCheckout(event: Event) {
+    if (event.target == null) { return; }
+
+    const orderId = osu.presence((event.target as HTMLElement).dataset.orderId);
+    if (orderId == null) {
+      throw new Error('orderId is missing');
+    }
+
+    const { isValid, lineItems } = this.collectShopifyItems();
+
+    if (!isValid) {
+      // can't mix Shopify and non-Shopify items.
+      // This should normally not show as the button itself shouldn't have been rendered.
+      osu.popup(osu.trans('model_validation/store/product.must_separate'), 'danger');
+
+      return;
+    }
+
+    if (lineItems.length > 0) {
+      return this.beginShopifyCheckout(orderId, lineItems);
+    }
+
+    Turbolinks.visit(laroute.route('store.checkout.show', { checkout: orderId }));
+  }
+
+  async beginShopifyCheckout(orderId: string, lineItems: LineItem[]) {
+    try {
+      LoadingOverlay.show();
+      LoadingOverlay.show.flush();
+
+      // create shopify checkout.
+      // error returned will be a JSON string in error.message
+      const checkout = await client.checkout.create({
+        customAttributes: [{ key: 'orderId', value: orderId }],
+        lineItems,
+      });
+
+      const params = {
+        orderId,
+        provider: 'shopify',
+        shopifyId: checkout.id,
+      };
+
+      await osu.promisify($.post(laroute.route('store.checkout.store'), params));
+
+      window.location = checkout.webUrl;
+    } catch (error) {
+      // either error from Shopify or updating the order state failed.
+      // TODO: separate the handling of errors.
+      osu.popup(osu.trans('errors.checkout.generic'), 'danger');
+      LoadingOverlay.hide();
+    }
+  }
+
+  async resumeCheckout(event: Event) {
+    if (event.target == null) { return; }
+
+    const target = event.target as HTMLElement;
+    const checkoutId = osu.presence(target.dataset.checkoutId);
+    if (checkoutId == null) {
```

Comment:
same as above, check provider name directly or in this case maybe call it `shopifyCheckoutId`?

...or transaction id.

Not sure if it makes sense having existence of checkout id means it's shopify checkout.